### DISCUSSION TITLE
TV Cast

### DISCUSSION THREAD

**Discussion by lapdv3012** [community/community#38739] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

Hi Bro, I'm currently making Cast app for TV & Browser combined ConnectSDK & Nanohttpd.
On Browser it works fine to cast mp3, mp4, & image files. But I don't understand why when casting on TV, only image files can work and mp3 mp4 files can't and if I copy the stream link to the browser, it still works. Anyone who has worked through this app, please teach me!.  Thank you!
[](https://github.com/NanoHttpd/nanohttpd)
[](https://connectsdk.com/en)
Here my code
`
package com.xxxxx.http;

import static com.xxxx;

import android.net.Uri;
import android.util.Log;
import android.webkit.MimeTypeMap;

import org.nanohttpd.protocols.http.IHTTPSession;
import org.nanohttpd.protocols.http.NanoHTTPD;
import org.nanohttpd.protocols.http.response.Response;
import org.nanohttpd.protocols.http.response.Status;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;

public class StreamingWebServer extends NanoHTTPD {

    private static StreamingWebServer instance;
    private static int port = 0;

    public int getPort() {
        return port;
    }

    public static synchronized StreamingWebServer getInstance() {
        if (null == instance) {
            if (port == 0) {
                port = nextFreePort(8000, 999);
            }
            instance = new StreamingWebServer(port);
        }
        return instance;
    }

    private StreamingWebServer(int port) {
        super(port);
        System.out.println("[[VideoHttpServer]] start");
    }

    @Override
    public Response handle(IHTTPSession session) {
        Map<String, String> headers = session.getHeaders();
        String uri = session.getUri();
        File f = new File(uri);
        System.out.println("[[VideoHttpServer]] handle" + f.getAbsolutePath());
        return serveFile(uri, headers, f);
    }

    private Response serveFile(String uri, Map<String, String> header, File file) {
        for (Map.Entry<String, String> entry : header.entrySet()) {
            String key = entry.getKey().toString();
            String value = entry.getValue();
            System.out.println(" VideoHttpServer key, " + key + " value " + value);
        }
        Response res;
        String mime = getMimeType(uri);

        System.out.println(" VideoHttpServer getMimeTypeForFile, " + mime);
        try {
            String etag = Integer.toHexString((file.getAbsolutePath() + file.lastModified() + "" + file.length()).hashCode());
            long startFrom = 0;
            long endAt = -1;
            String range = header.get("range");
            if (range != null) {
                if (range.startsWith("bytes=")) {
                    range = range.substring("bytes=".length());
                    int minus = range.indexOf('-');
                    try {
                        if (minus > 0) {
                            startFrom = Long.parseLong(range.substring(0, minus));
                            endAt = Long.parseLong(range.substring(minus + 1));
                        }
                    } catch (NumberFormatException ignored) {
                    }
                }
            }
            long fileLen = file.length();
            if (range != null && startFrom >= 0) {
                if (startFrom >= fileLen) {
                    res = createResponse(Status.RANGE_NOT_SATISFIABLE, MIME_PLAINTEXT, "");
                    res.addHeader("Content-Range", "bytes 0-0/" + fileLen);
                    res.addHeader("ETag", etag);
                } else {
                    if (endAt < 0) {
                        endAt = fileLen - 1;
                    }
                    //endAt=startFrom+1000000;
                    long newLen = (endAt - startFrom) + 1;
                    if (newLen < 0) {
                        newLen = 0;
                    }

                    final long dataLen = newLen;
                    FileInputStream fis = new FileInputStream(file) {
                        @Override
                        public int available() throws IOException {
                            return (int) dataLen;
                        }
                    };
                    fis.skip(startFrom);

                    res = createResponse(Status.PARTIAL_CONTENT, mime, fis, dataLen);
                    res.addHeader("Content-Length", "" + dataLen);
                    res.addHeader("Content-Range", "bytes " + startFrom + "-" + endAt + "/" + fileLen);
                    res.addHeader("ETag", etag);
                    Log.d("VideoHttpServer", "serveFile --1--: Start:" + startFrom + " End:" + endAt);
                }
            } else {
                if (etag.equals(header.get("if-none-match"))) {
                    res = createResponse(Status.NOT_MODIFIED, mime, "");
                    Log.d("VideoHttpServer", "serveFile --2--: Start:" + startFrom + " End:" + endAt);
                } else {
                    FileInputStream fis = new FileInputStream(file);
                    res = createResponse(Status.OK, mime, fis, fis.available());
                    res.addHeader("Content-Length", "" + fileLen);
                    res.addHeader("ETag", etag);
                    Log.d("VideoHttpServer", "serveFile --3--: Start:" + startFrom + " End:" + endAt);
                }
            }
        } catch (IOException ioe) {
            res = getResponse("Forbidden: Reading file failed");
        }

        return (res == null) ? getResponse("Error 404: File not found") : res;
    }

    private Response createResponse(Status status, String mimeType, InputStream inputStream, long lenght) {
        Response newFixedLengthResponse = Response.newFixedLengthResponse(status, mimeType, inputStream, lenght);
        newFixedLengthResponse.addHeader("Accept-Ranges", "bytes");
        return newFixedLengthResponse;
    }

    private Response createResponse(Status status, String mimeType, String txt) {
        Response newFixedLengthResponse = Response.newFixedLengthResponse(status, mimeType, txt);
        newFixedLengthResponse.addHeader("Accept-Ranges", "bytes");
        return newFixedLengthResponse;
    }

    private Response getResponse(String txt) {
        return createResponse(Status.OK, "text/plain", txt);
    }

    public static String getMimeType(String uri) {
        String fileExtensionFromUrl = MimeTypeMap.getFileExtensionFromUrl(sanitizeFileName(Uri.encode(uri)));
        if (fileExtensionFromUrl != null) {
            return MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtensionFromUrl);
        }
        return null;
    }

    public static String sanitizeFileName(String str) {
        byte[] bArr = {34, 60, 62, 124, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 58, 42, 63, 92, 47};
        for (int i = 0; i < 41; i++) {
            str = str.replace((char) bArr[i], '_');
        }
        return str;
    }
}

`

**Comment by ossanna16** [comment]
Hi @lapdv3012,
Thanks for being a part of the GitHub Community! We're glad you're here.

I wanted to jump in and mention that if you're looking for help for this specific topic, you might want to try asking for help somewhere that focuses on TV Casts. It's definitely possible that another GitHub user might have run into this same issue and can help, but the GitHub Community on Discussions focuses primarily on topics related to GitHub itself or collaboration on project development and ideas. We want to make sure you’re getting the best support you can, but this space may not be the right place for this particular topic.

Best of luck!
